tardis/io/util.py

```diff
@@ -3,12 +3,127 @@
+    Parameters
+    ----------
+    text:
+        The string to convert to `astropy.units.Quantity`
+    Returns
+    -------
+    `astropy.units.Quantity`
+    """
+    value_str, unit = text.split(None, 1)
+    value = float(value_str)
+    if unit.strip() == 'log_lsun':
+        value = 10 ** (value + np.log10(constants.L_sun.cgs.value))
+        unit = 'erg/s'
+    return u.Quantity(value, unit)
+
+
+class MockRegexPattern(object):
+    """
+    A mock class to be used in place of a compiled regular expression
+    when a type check is needed instead of a regex match.
+
+    Note: This is usually a lot slower than regex matching.
+    """
+    def __init__(self, target_type):
+        self.type = target_type
+
+    def match(self, text):
+        """
+
+        Parameters
+        ----------
+        text:
+            A string to be passed to `target_type` for conversion.
+        Returns
+        -------
+        `True` if `text` can be converted to `target_type`.
+        """
+        try:
+            self.type(text)
+        except ValueError:
+            return False
+        return True
+
+
+class YAMLLoader(yaml.Loader):
+    """
+    A custom YAML loader containing all the constructors required
+    to properly parse the tardis configuration.
+    """
+    @classmethod
+    def add_implicit_resolver(cls, tag, regexp, first=None):
+        """
+        Parameters
+        ----------
+
+        tag:
+            The YAML tag to implicitly apply to any YAML scalar that matches `regexp`
+
+        regexp:
+            The regular expression to match YAML scalars against `tag`
+
+
+        Notes
+        -----
+
+        This classmethod is a monkey-patch for a copy() related bug
+        in the original class method which affects this yaml.Loader subclass.
+
+        This class method is to be removed when this bug gets fixed upstream.
+
+        https://bitbucket.org/xi/pyyaml/issues/57/add_implicit_resolver-on-a-subclass-may
+        """
+        if 'yaml_implicit_resolvers' not in cls.__dict__:
+            yaml_implicit_resolvers = {}
+            for k, v in cls.yaml_implicit_resolvers.items():
+                yaml_implicit_resolvers[k] = copy.copy(v)
+            cls.yaml_implicit_resolvers = yaml_implicit_resolvers
+        if first is None:
+            first = [None]
+        for ch in first:
+            cls.yaml_implicit_resolvers.setdefault(ch, []).append((tag, regexp))
+
+    def construct_quantity(self, node):
+        """
+        A constructor for converting quantity-like YAML nodes to
+        `astropy.units.Quantity` objects.
+
+        Parameters
+        ----------
+
+        node:
+            The YAML node to be constructed
+
+        Returns
+        -------
+
+        `astropy.units.Quantity`
+
+        """
+        data = self.construct_scalar(node)
+        return quantity_from_str(data)
+
+YAMLLoader.add_constructor(u'!quantity', YAMLLoader.construct_quantity)
```

Comment:
Not really, this is the proper way to do it. `__new__` is used to create a new instance (and `__init__` to initialize it). A senseful alternative to how I currently do it would be something like:

```python
class YAMLLoader(yaml.Loader):
    yaml_implicit_resolvers = yaml.Loader.yaml_implicit_resolvers + new_implicit_resolvers # Pseudocode
    yaml_constructors = yaml.Loader.yaml_constructors + new_constructors # Pseudocode
```
but this is not what I chose to do because there are some extra things that `add_implicit_resolver` takes care of and because this is why it is there for. 

This is a common thing when you need to change a class. If you choose to call these classmethods on instance creation/initialization you'll find yourself adding the same stuff multiple times since they actually change the class attributes, which are shared between all instances. Then you'd have to do a check like `if my_new_resolver in YAMLLoader.yaml_implicit_resolvers then don't_add_it_again` which doesn't make any sense to do.

This is also the reason I created a separate class. I could simply do `yaml.Loader.add_constructor/resolver()` but then anything that would use `yaml.load()` from this point and on would be forced to have my customizations.

If pyyaml had the resolvers/constructors as instance attributes instead of class attributes there would be no problem in doing what you said, or even
```
myloader = yaml.Loader(...)
myloader.add_constructor/resolver(...)
```
without ever needing to subclass `yaml.Loader`

Lastly, this is how even pyyaml itself does it for its default Resolver class:

```python
class BaseResolver(object):

    DEFAULT_SCALAR_TAG = u'tag:yaml.org,2002:str'
    DEFAULT_SEQUENCE_TAG = u'tag:yaml.org,2002:seq'
    DEFAULT_MAPPING_TAG = u'tag:yaml.org,2002:map'

    yaml_implicit_resolvers = {}
    yaml_path_resolvers = {}

    def __init__(self):
        self.resolver_exact_paths = []
        self.resolver_prefix_paths = []
(snip)
class Resolver(BaseResolver):
    pass

Resolver.add_implicit_resolver(
        u'tag:yaml.org,2002:bool',
        re.compile(ur'''^(?:yes|Yes|YES|no|No|NO
                    |true|True|TRUE|false|False|FALSE
                    |on|On|ON|off|Off|OFF)$''', re.X),
        list(u'yYnNtTfFoO'))

Resolver.add_implicit_resolver(
        u'tag:yaml.org,2002:float',
        re.compile(ur'''^(?:[-+]?(?:[0-9][0-9_]*)\.[0-9_]*(?:[eE][-+][0-9]+)?
                    |\.[0-9_]+(?:[eE][-+][0-9]+)?
                    |[-+]?[0-9][0-9_]*(?::[0-5]?[0-9])+\.[0-9_]*
                    |[-+]?\.(?:inf|Inf|INF)
                    |\.(?:nan|NaN|NAN))$''', re.X),
        list(u'-+0123456789.'))
```